proposals/explicit-backing-fields.md

```diff
@@ -80,100 +109,155 @@ val loading: LiveData<Boolean>
->  - This is another place where Java pattern could be introduced into Kotlin with less boilerplate
-
-The proposed syntax allows to achieve the same functionality while keeping the same level of simplicity:
+In this use-case, an read access to the field from inside the corresponding classes/modules 
+(where the field is visible) automatically gives access to mutable type, which we call [Smart Type Narrowing](#smart-type-narrowing), 
+but is seen as a property with read-only type to the outside code.
+
+### Decouple storage type from external representation
+
+Sometimes a property must be internally represented by a different type for storage-efficiency or architectural reason,
+while having a different outside type. For example, API requirements might dictate that the property type
+is `String`, but if we know that it always represents a decimal integer, then it can be efficiently stored as such 
+with custom getter and custom setter.
 
 ```kotlin
-class MyClass {
-    val data: List<String>
-        field = ArrayList()
-}
+val number: String
+    field: Int = 0
+    get() = field.toString()
+    set(value) { field = value.toInt() }
 ```
 
-### RX Observable and Subjects
+### Expose read-only view
+
+In some application it is desirable to expose not just a read-only subtype, but a specially constructed read-only
+view that protects the data structure from casting into mutable type. For example a `MutableStateFlow` has
+[`asStateFlow`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/as-state-flow.html)
+extension for that purpose.
 
-> Data can often be pushed into reactive world using subjects. However, exposing `Subject` would allow consumer of the class to push its own data into it. That is why it is good idea to expose all subjects as read-only `Observable`:
+With the proposed syntax it is possible to declare a custom getter:
 
 ```kotlin
-class MyClass {
-    private val _dataStream = PublishSubject.create<String>()
-    val dataStream: Observable<String>
-        get() = _dataStream
-}
+val state: StateFlow<State> 
+    get() = field.asStateFlow()
+    field = MutableStateFlow(State.INITIAL)
 ```
 
-Turns into:
+For this use-case, the TBD syntax of [Direct Backing Field Access](#direct-backing-field-access) will need to be added.
+
+### Access field from outside of getter and setter
+
+Kotlin allows property field to be accessed from the property's getter and setter using a `field` variable.
+This proposal is designed with an idea to provide an explicit syntax to access a property's field from anywhere
+inside the corresponding class when the field is declared explicitly:
 
 ```kotlin
-class MyClass {
-    val dataStream: Observable<String>
-        field = PublishSubject.create()
+class Component {
+    var status: Status
+        field // explicit field declaration
```

Comment:
This line is redundant as it is

Reply:
Also, there is no initializer here

Reply:
Let's discuss this part of design here. Do we automatically make backing fields of _all properties_ available inside the class via TBD syntax or only if you explicitly declare a backing field for them?

Reply:
I think this is better

> we automatically make backing fields of all properties available inside the class via TBD syntax

Reply:
Ok. Corrected.